### ISSUE TITLE
[Afya Yangu/Quest] - Family Registration

### ISSUE THREAD

**Issue by samkanga** [opensrp/fhircore#1264]
**Is your feature request related to a problem? Please describe.**
Afya Yangu is the new name for Boresha Afya project that is being transitioned to FHIR. This ticket forms the first of a couple of modules that need to be built in the realisation of the Afya Yangu CHW App.

**Describe the solution you'd like**
As a Community Health Worker (CHW) I need to register households and add household members while at the community conducting visits. Household and household members registrations are made once and they are then referenced while providing the different services at the community. Household members can either be children under 5 years of age or older people. One household can contain multiple household members.

- [ ] I would like to view all my registered households, under a common section “All Families”. 
- [ ] Each of the listed households should provide a summary with family name, the village where the family is located, icons to show the number of household members of programmatic interest. These include children under 5, women under ANC, PNC or FP, and individuals with a Malaria diagnosis. 
- [ ] Each of the families should provide a count of all the services that are due (coded in blue) or overdue (coded in red).
- [ ] I would also like to have an option to add a new household whenever I need to. Once a new household is added, it should immediately be available under the “All Families” listing. 
- [ ] I would like to have an option for searching for the families/households by their name or system ID.
- [ ] I would like to have the ability to filter the household listing by those that have due services

The household registration questionnaire can be accessed [here](https://github.com/opensrp/fhir-resources/blob/main/afya_yangu/questionnaires/household_registration.json).

When I select a family: 

- [ ] I should see the listing of all the existing family members. 
- [ ] For each of the listed family members I should see a summary of the person’s full name, sex, age and whether that person is the head of the household or the primary care giver. 
- [ ] From the household members listing I should be able to know family members who have pending services or who I have already provided with the services due to them. Services include home visits for under 5 and PNC mothers among others. This can be through a coloured bubble next to the household member (blue where services are given and red where the services are due).
- [ ] I should be able to select a household member from the household members listing and view their information in detail. i.e. Name, Age, Sex, Village, ID and a listing of the services that are due to the person or to his/her family.
- [ ] The household member listing should have capability to divide the listing into pages where there are many household member.
- [ ] I should have the option to change who is the household head or the primary care giver.

From within a household:

- [ ] I should have the option to add a new household member. 
- [ ] At the point of adding a new household member, I want separate options for adding a child under 5 and for adding an older child or person. This is because part of the information collected between the two groups of people differs.
- [ ] I should have an option to initiate a call to the household member marked as the head of the household.
- [ ] I should be able to remove a household member from the family when such a person moves away, dies or is no longer part of the household for any reason. The App should provide me with the listing of all the members in the selected household. 
- [ ] I should then have the capability to select the household member to be removed and provide the reason for removal. 
- [ ] The removed person's record should not be accessible after removal. No other step needs to be taken after the removal from a household.
- [ ] I should be able to remove a household. Removal of a household will also result in the removal of all the household members associated with it.
- [ ] The removed household should not be listed under the "All Families" list and should not return results from a search by the family name or ID.

Below are the links to additional questionnaires used in the family registration.

1.  [Adding Households](https://github.com/opensrp/fhir-resources/blob/main/afya_yangu/questionnaires/household_registration.json)
2. [Adding children under 5](https://github.com/opensrp/fhir-resources/blob/main/afya_yangu/questionnaires/under-5-registration.fhir.json)
3. [Adding HH members above 5 years of age](https://github.com/opensrp/fhir-resources/blob/main/afya_yangu/questionnaires/other-family-member-registration.fhir.json)
4. [Removing HH Member](https://github.com/opensrp/fhir-resources/blob/main/afya_yangu/questionnaires/remove-family-member.json)
5. [Removing Family Record](https://github.com/opensrp/fhir-resources/blob/main/afya_yangu/questionnaires/remove-family-record.json)

**Describe alternatives you've considered**
N/A

**Additional context**

**Comment by pld** [comment]
What in this is not already supported?

**Comment by f-odhiambo** [comment]
@pld A majority of the functionality is already available just a few edge cases such as 

1. _At the point of adding a new household member, I want separate options for adding a child under 5 and for adding an older child or person. This is because part of the information collected between the two groups of people differs._
2. Addition of linked StructureMaps for the authored questionnaires 
3. 3. Acceptance criteria for QA  - TBD @samkanga to add 

We can  keep this ticket to track both code  and non-code items for configurability 

CC @samkanga

**Comment by pld** [comment]
Cool, I'm closing this ticket, we can use the closed ticket as a reference. Even for (1) I think a Questionnaire can handle that.

We can open a new ticket when any missing functionality arises that requires code and clear describes the code that must be written to close it.

For the content issues, we can track those elsewhere, not tickets in this code repo.

**Comment by f-odhiambo** [comment]
@samkanga and @Gental-Giant to update the ticket accordingly

**Comment by samkanga** [comment]
@f-odhiambo cc: @Gental-Giant @ageryck  See the updates below:

1. We merged the under 5 and over 5 registration forms and have since removed the questions that are not related to the family member registration such as whether there was a delivery registered in the last 1 year or if a child has a growth booklet.
2. The request to provide a separation of the registration forms is no-longer needed and has been crossed out.
3. See additional detail below:

- Updated Afya Yangu (AY) questionnaires on server as below; 
- [Household](https://fhir.labs.smartregister.org/fhir/Questionnaire/128804) - Updated
- [Family Member Registration](https://fhir.labs.smartregister.org/fhir/Questionnaire/128806) - Updated
- [Remove House Hold](https://fhir.labs.smartregister.org/fhir/Questionnaire/128807) - No Changes
- [Remove Family Member](https://fhir.labs.smartregister.org/fhir/Questionnaire/128808) - Updated
- [Child U5](https://fhir.labs.smartregister.org/fhir/Questionnaire/128805) - To discard, no longer in use

**Comment by f-odhiambo** [comment]
For the Questionnaire will need to have the following subsequent changes 
1. change the subject type to Group from Patient.
2. add group definition and extension in Family Name QuetionnaireItem.
3. add household ID in the questionnaire.

**Comment by ageryck** [comment]
@f-odhiambo changes above done. Is `Group.identifier` a requirement too ?

**Comment by dubdabasoduba** [comment]
> @pld A majority of the functionality is already available just a few edge cases such as
> 
> 1. _At the point of adding a new household member, I want separate options for adding a child under 5 and for adding an older child or person. This is because part of the information collected between the two groups of people differs._
> 2. Addition of linked StructureMaps for the authored questionnaires
> 3. Acceptance criteria for QA  - TBD @samkanga to add
> 
> We can keep this ticket to track both code and non-code items for configurability
> 
> CC @samkanga

Did we already create a ticket for the seperation? Can we use the enableWhen logic on the forms to show and hide the questions based on who is being registered?

- Have the first question be _`What type of client do you want to register?`_ This question can have the option of under 5 and over 5. We then hide or show the other questions based on this.

**Comment by ageryck** [comment]
@dubdabasoduba that would be our last resort(we discussed this already) should we be unable to use runtime calculated `age ` variable from `DOB ` to implement the skip logic. However even in this option we would still need to restrict DOB entry values based on responded value i.e. for U5 the calendar must not allow any entry above 5 years.

**Comment by dubdabasoduba** [comment]
> @dubdabasoduba that would be our last resort(we discussed this already) should we be unable to use runtime calculated `age ` variable from `DOB ` to implement the skip logic. However, even in this option, we would still need to restrict DOB entry values based on responded value i.e. for U5 the calendar must not allow any entry above 5 years.

I don't think we have a function to calculate DOB at runtime. FHIRPath which is used for extraction doesn't have runtime support for such calculations. 

I think the date restrictions can be added by the min & max settings on the date field. We can test that to see if it works well.

**Comment by f-odhiambo** [comment]
Just sharing some related tickets on the date issue 
1. https://github.com/google/android-fhir/issues/881
2. https://github.com/google/android-fhir/issues/1040

Also, related items shared by @ageryck  - Support simple calculation of patient age in fhirpath [#1014](https://github.com/IBM/FHIR/issues/1014 
)

CC @ageryck @samkanga @dubdabasoduba

**Comment by Gental-Giant** [comment]
> @ageryck @samkanga I just went through the issue again. Could we update this to different issues? Some of the points are more about scheduling and tasking and not registration. We can have a call to do this.

@dubdabasoduba Sure we can do that.

**Comment by dubdabasoduba** [comment]
> Reviewing some items here
> 
> 1. Do we need Is family head? choice? where will this be used
> 
> The following items seem to be extensions of the patient
> 
> 1. Is the primary caregiver
> 2. Physical disabilities
> 3. Type of physical disabilities? if the Patient has Physical disabilities?
> 4. Occupation
> 5. Leadership role in the community
> 
> These seem like things that will need to be discussed as they may or may not be available as Patient extensions else the data captured can be left stored in the QuestionnaireResponse
> 
> CC @samkanga @ageryck @dubdabasoduba

@f-odhiambo Can we try to review whether we can use the encounters for this? This will mean the product needs to address the following things. 
- Pre-populating a Questionnaire from 2 resources. 
- Make sure that every Questionnaire filled up has a corresponding encounter generated for it.

**Comment by dubdabasoduba** [comment]
> > > @dubdabasoduba that would be our last resort(we discussed this already) should we be unable to use runtime calculated `age ` variable from `DOB ` to implement the skip logic. However, even in this option, we would still need to restrict DOB entry values based on responded value i.e. for U5 the calendar must not allow any entry above 5 years.
> > 
> > 
> > I don't think we have a function to calculate DOB at runtime. FHIRPath which is used for extraction doesn't have runtime support for such calculations.
> > I think the date restrictions can be added by the min & max settings on the date field. We can test that to see if it works well.
> 
> Won't this issue address this? [google/android-fhir#1040](https://github.com/google/android-fhir/issues/1040)

@pld I don't this issue address the AGE  from DOB and vice versa on runtime. This is more on adding the min/max restrictions on the date field.

**Comment by pld** [comment]
True, but, if I'm reading this correctly the only reason we care about calculating the age in the form is so we can have skip logic for if the person if <5, if 1040 is closed we can should be able to use today() to do that calculation, like today()-5 years > dob

**Comment by dubdabasoduba** [comment]
My understanding is this is to allow restrictions like making sure that someone's DOB can only be <= today() and not in the future. I don't think it adds the ability to evaluate `today() - 5 years > dob` 

More context on the issue is found here https://github.com/opensrp/fhircore/issues/845

**Comment by pld** [comment]
Yea good point, that issue is only relevant for the validation. I guess I was figuring that if that issue shows that today() is support in FHIRPath, and we know that FHIRPath already supports date arithmetic for adding and subtracting durations to a date, https://hl7.org/fhirpath/#datetime-arithmetic, we could use that elsewhere in a Questionnaire to evaluate today() - 5 years > dob

**Comment by dubdabasoduba** [comment]
> Yea good point, that issue is only relevant for the validation. I guess I was figuring that if that issue shows that today() is support in FHIRPath, and we know that FHIRPath already supports date arithmetic for adding and subtracting durations to a date, https://hl7.org/fhirpath/#datetime-arithmetic, we could use that elsewhere in a Questionnaire to evaluate today() - 5 years > dob

I don't think it would during runtime though. We need to test it better. it could work during extraction I think we have tested a couple of date calculations.

**Comment by pld** [comment]
Does the validation execute during run-time?

**Comment by RaaziaTarique** [comment]
The following fields are left that needs to be mapped from family/household registration questionnaire. I am unable to find fields in the group resource that can be used.

-   `Village: string`
-   `Landmark/ description of house location:  string`
-   `Nearest health facility:  string`

CC: @f-odhiambo @shoaibmushtaq25 @maimoonak
